phabricatoremails/template.py

```diff
@@ -0,0 +1,237 @@
+import textwrap
+from datetime import datetime, timezone
+
+import jinja2
+from phabricatoremails.model.common import Reviewer, ReviewerStatus
+from phabricatoremails.model.phabricator import (
+    ReplyContext,
+    CodeContext,
+    DiffLineType,
+    AffectedFileChange,
+)
+from phabricatoremails.model.phabricator_secure import ExistenceChange
+from premailer import Premailer
+
+
+def is_reply(context):
+    return isinstance(context, ReplyContext)
+
+
+def is_code(context):
+    return isinstance(context, CodeContext)
+
+
+def is_has_comments(event):
+    return event.main_comment or event.inline_comments
+
+
+def is_has_comments_secure(event):
+    return event.comment_count > 0
+
+
+def is_reviewer_accepted(reviewer: Reviewer):
+    return not reviewer.is_actionable and reviewer.status == ReviewerStatus.ACCEPTED
+
+
+def is_reviewer_blocking(reviewer: Reviewer):
+    return reviewer.status == ReviewerStatus.BLOCKING
+
+
+def is_reviewer_soft_needed(reviewer: Reviewer):
+    # The revision needs a review from _someone_, and this person _is_ a reviewer, but
+    # it's not them specifically who needs to perform the review
+    return reviewer.is_actionable and reviewer.status == ReviewerStatus.UNREVIEWED
+
+
+def is_reviewer_needed(reviewer: Reviewer):
+    # This specific reviewer needs to accept this revision before it can be landed
+    return reviewer.is_actionable and reviewer.status in [
+        ReviewerStatus.REQUESTED_CHANGES,
+        ReviewerStatus.BLOCKING,
+    ]
+
+
+def reviewer_actionable_class(reviewer: Reviewer):
+    return "actionable" if reviewer.is_actionable else "non-actionable"
+
+
+def date(utc_value: datetime, receiver_timezone: timezone):
+    return utc_value.astimezone(receiver_timezone).strftime("%b %-d %-I:%M%p:")
```

Comment:
Factor our the time template to its own variable and give it a meaningful name (or documentation).

Reply:
Good call on adding some docs here explaining the template, but why move it to its own variable?

Reply:
Putting it in a variable will allow you to self-document what the variable is for. Your method should have a docstring saying what the method does.

It's conceivable in the future that this template could be used elsewhere, for example in a test.

Reply:
> Putting it in a variable will allow you to self-document what the variable is for.

An inline comment would provide the same benefit, right?

> It's conceivable in the future that this template could be used elsewhere, for example in a test.

This is true, but we can expose this in a variable when it's used elsewhere.

Reply:
This would fall under the general "bad hard coding practices", similar to what I mentioned in my comment regarding line length.

Unless there's a good reason for it, hard coded values should be put into a variable and given a meaningful name.

Reply:
> This would fall under the general "bad hard coding practices", similar to what I mentioned in my comment regarding line length.

Can you link this comment? I can't find it in this PR :thinking: 

> Unless there's a good reason for it, hard coded values should be put into a variable and given a meaningful name.

I understand two benefits to putting a hard coded value into a variable:
1. You can document the value with the variable name
2. If the value is used elsewhere, you can updated it in a central location

Point 1 is covered by inline documentation, and point 2 is covered because the value isn't used elsewhere (yet). Is there another benefit to using a constant that I'm missing?